CHUI-69: Fix issues in first purchase flow

marcoskwkm opened the pull request:
#### What problem is this solving?

This PR fixes some of the UX issues identified by the designers:
1. The address is displayed in minimal mode (postal code only) on the SLA selection step to match the [specification](https://www.figma.com/proto/O84mDk0F6zC3F06unCp1iJLk/01.-Cart-and-Checkout-Mockups?node-id=9888%3A405541&viewport=48%2C-136%2C0.4082198441028595&scaling=min-zoom)
2. After editing the receiver name, the user is redirected to the payment step instead of the SLA selection step.
3. The user cannot get past the complete address step if there are invalid fields.
4. Clicking the edit address button on the SLA selection screen now takes the user to different steps depending on certain conditions.

This is still missing the fix to the inconsistency of the displayed address input fields between the first time the user reaches the complete address step and the next times. I'll take care of that in following PRs because that might require a slightly more elaborate solution.

*Note:* This is rebased on `feature/address-receiver-name` and depends on https://github.com/vtex-apps/address-context/pull/6.

#### How should this be manually tested?

1. [Add an item to cart.](https://shippingflow--checkoutio.myvtex.com/cart/add?sku=289)
2. Proceed to checkout and use a first-purchase email.
3. In the shipping step, use `22250-040` as postal code. The SLA selection screen should show only the postal code you entered.
4. Click the edit address button. You should be taken back to the postal code input. Redo step 3.
5. Select any SLA. In the complete address screen, click continue without filling the number input. Nothing should happen.
6. Fill the number field, uncheck the "will receive the order" checkbox and click continue.
7. Write any receiver name and click continue. You should be taken to the payment step.
8. Return to the shipping step and click the edit address button. You should be taken to the complete address form.

#### Checklist/Reminders

- [ ] Updated `README.md`.
- [x] Updated `CHANGELOG.md`.
- [ ] Linked this PR to a Clubhouse story (if applicable).
- [ ] Updated/created tests (important for bug fixes).
- [ ] Deleted the workspace after merging this PR (if applicable).

#### Screenshots or example usage

#### Type of changes

<!--- Add a ✔️ where applicable -->
✔️ | Type of Change
---|---
✔️ | Bug fix <!-- a non-breaking change which fixes an issue -->
_ | New feature <!-- a non-breaking change which adds functionality -->
_ | Breaking change <!-- fix or feature that would cause existing functionality to change -->
_ | Technical improvements <!-- chores, refactors and overall reduction of technical debt -->

Commit:
Use minimal display for a new incomplete address in SLA selection

Commit:
Send user to payment after editing receiver name

Commit:
Prevent user from advancing from complete address stage if address is invalid

Commit:
Update changelog

lucasecdb commented on react/shippingStateMachine/shippingStateMachine.ts:
```diff
@@ -141,7 +141,7 @@ const shippingStateMachine = Machine<
               src: 'tryToEditReceiverInfo',
               onDone: {
-                target: '#shipping.selectDeliveryOption',
+                target: '#shipping.done',
```
this might be intended but the flow to edit the receiver name on the address would always redirect the user to the payment step?

in case you went back to the shipping just to edit the receiver name (the flow being: other step -> edit shipping -> sla selection list -> edit receiver name -> payment) would it be desirable to redirect the user back to the SLA selection list or to send the user to payment? cc @augustocb

marcoskwkm replied:
I discussed with @augustocb and the intended flow will always be to send the user to the payment step, because the address is already completed and there is no need for more actions from the user.

lucasecdb commented on react/components/ShippingOptionList.tsx:
```diff
@@ -64,7 +64,7 @@ const ShippingOptionList: React.FC<Props> = ({
 
       <div className="mt2 mb6 lh-copy">
-        <PlaceDetails display="compact" />
+        <PlaceDetails display="minimal" />
```
this would cause some weirdness when selecting an address from the address list and only see the zipcode in the SLA selection list, right?

marcoskwkm replied:
Good point. We will use the minimal display when the user enters a new address, which is always the case in the first purchase scenario. In second purchase scenarios this needs to be adjusted, but I guess I'll take care of that in a following PR.

marcoskwkm replied:
Right. I changed it to only use the `minimal` display when the address is complete, which for now I'm going to assume is equivalent to `isValid === true`. This doesn't work in the identified but not authenticated scenario (`isValid` becomes `false`), but considering this PR focuses on the first purchase flow I guess we can take care of that in future PRs.

lucasecdb replied:
sounds good!

marcoskwkm commented:
Converting to draft while I wait for a review of this flow from the designers.

marcoskwkm reviewed (commented):
~Reopening for review as the design pendencies were resolved.~
EDIT: Nevermind.

Commit:
Send user to appropriate step when editing address

Commit:
Fix spacing in address completion form

marcoskwkm commented:
Ok, should be all good now.

augustocb commented:
When getting back from Payment to Shipping and editing address, should be shown:
- the number field should be shown
- info about the receiver

![image](https://user-images.githubusercontent.com/1315451/86784854-5e3d5200-c038-11ea-92ab-78648e6dd242.png)

lucasecdb commented:
for some reason, when I start typing in the number field, the checkbox for the receiver name disappears

![pr-69](https://user-images.githubusercontent.com/10223856/86785621-4b774d00-c039-11ea-9dd0-2f60d564ad8f.gif)

edit: it only happens when you click the submit button _before_ editing the number field

lucasecdb commented on react/components/AddressCompletionForm.tsx:
```diff
@@ -62,7 +62,12 @@ const AddressCompletionForm: React.FC<Props> = ({
       updatedAddress.receiverName = `${firstName} ${lastName}`
```
the issue above is because this line is mutating the address object in the context.. I think it could be solved if we cloned the object instead of just assigning it to the `updatedAddress` variable.

marcoskwkm replied:
This does not explain why the checkbox disappears when you type something in the input, right? Because the form is not being submitted at that time. I only managed to reproduce this once, and I don't even know how I did it...
Anyway, I changed this component so that the condition to show the checkbox uses the `selectedAddress` from the state. I think that should work as well.

lucasecdb replied:
I could reproduce it consistently by pressing the submit button and then type something in the input. It kinda makes sense if you see that we are mutating the object that is being stored in the `AddressContextProvider` state.

So, when we type in the number field input and the `AddressForm` component updates the address with the new number field value ([here](https://github.com/vtex-apps/place-components/blob/master/react/AddressForm.tsx#L218-L221)), the `prevAddress` already has the receiver name because we mutated it, and because it is a `setState`, the `AddressCompletionForm` re-renders and the check `receiverName != null` is now `false`, which causes us to remove the checkbox

marcoskwkm replied:
Ah! Naruhodo.
I'll clone the `address` object here then to prevent mutating the context state object and probably causing more weird bugs in the future.
EDIT: Done.

Commit:
Show receiver info on complete address screen when available

marcoskwkm commented:
> When getting back from Payment to Shipping and editing address, should be shown:
> 
> * the number field should be shown

This is a bit trickier so I'll do this in a separate PR.

> * info about the receiver

Done. Could you check once more?

Commit:
Clone address object to avoid mutating the context state

lucasecdb reviewed (approved):
LGTM!

Commit:
Update local selectedAddress after completing address

marcoskwkm commented:
- Rebased
- Added a commit that updates the machine context's `selectedAddress` after the complete address step. Honestly, I'm not sure how this was working before.

jeffersontuc commented:
Nice job!

lucasecdb commented:
>   * Added a commit that updates the machine context's `selectedAddress` after the complete address step. Honestly, I'm not sure how this was working before.

I think I did the same in #8 😅

marcoskwkm commented:
Rebased and removed typings from local workspace.
